crates/matrix-sdk-crypto-nodejs/src/machine.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+#[napi]
+impl OlmMachine {
+    // napi doesn't support `#[napi(factory)]` with an `async fn`. So
+    // we create a normal `async fn` function, and then, we create a
+    // constructor that raises an error.
+
+    #[napi]
+    pub async fn initialize(
+        user_id: &identifiers::UserId,
+        device_id: &identifiers::DeviceId,
+    ) -> Self {
+        let user_id = user_id.inner.clone();
+        let device_id = device_id.inner.clone();
+
+        OlmMachine {
+            inner: matrix_sdk_crypto::OlmMachine::new(user_id.as_ref(), device_id.as_ref()).await,
+        }
+    }
+
+    #[napi(constructor)]
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> napi::Error {
+        napi::Error::from_reason("To build an `OldMachine`, please use the `initialize` method")
+    }
+
+    #[napi]
+    pub fn user_id(&self) -> identifiers::UserId {
+        identifiers::UserId::new_with(self.inner.user_id().to_owned())
+    }
+
+    #[napi]
+    pub fn device_id(&self) -> identifiers::DeviceId {
+        identifiers::DeviceId::new_with(self.inner.device_id().to_owned())
+    }
+
+    #[napi]
+    pub fn identity_keys(&self) -> IdentityKeys {
+        self.inner.identity_keys().into()
+    }
+
+    #[napi]
+    pub async fn receive_sync_changes(
+        &self,
+        to_device_events: String,
+        changed_devices: &sync_events::DeviceLists,
+        one_time_key_counts: HashMap<String, u32>,
+        unused_fallback_keys: Vec<String>,
+    ) -> Result<String, napi::Error> {
+        let to_device_events = serde_json::from_str(to_device_events.as_ref()).map_err(into_err)?;
+        let changed_devices = changed_devices.inner.clone();
+        let one_time_key_counts = one_time_key_counts
+            .iter()
+            .filter_map(|(key, value)| {
+                Some((DeviceKeyAlgorithm::from(key.as_str()), UInt::new(*value as u64)?))
+            })
+            .collect::<BTreeMap<DeviceKeyAlgorithm, UInt>>();
+        let unused_fallback_keys = Some(
+            unused_fallback_keys
+                .into_iter()
+                .map(|key| DeviceKeyAlgorithm::from(key.as_str()))
+                .collect::<Vec<DeviceKeyAlgorithm>>(),
+        );
+
+        Ok(serde_json::to_string(
+            &self
+                .inner
+                .receive_sync_changes(
+                    to_device_events,
+                    &changed_devices,
+                    &one_time_key_counts,
+                    unused_fallback_keys.as_deref(),
+                )
+                .await
+                .map_err(into_err)?,
+        )
+        .map_err(into_err)?)
+    }
+
+    // We could be tempted to use `requests::OutgoingRequests` as its
+    // a type alias for this giant `Either` chain. But `napi` won't
+    // unfold it properly into a valid TypeScript definition, so…
+    // let's copy-paste :-(.
+    #[napi]
+    pub async fn outgoing_requests(
+        &self,
+    ) -> Result<
+        Vec<
+            Either<
```

Comment:
Can we at least type alias this monstrosity?

Reply:
See comment above the function..

Reply:
Ah missed that it was talking about a type alias, considering that `OutgoingRequests` is not a type alias.